### PR TITLE
Switching to logr tag v1.0.0

### PR DESCRIPTION
**Pull request by dims**
This is a **BREAKING** change. We should probably switch to klog to v3 as it changes the interface :(

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: klog.go [hunk 1]
```diff
@@ -509,7 +509,7 @@ type loggingT struct {
 	addDirHeader bool
 
 	// If set, all output will be redirected unconditionally to the provided logr.Logger
-	logr logr.Logger
+	logr *logr.Logger
```
Comment (pohly) [inline klog.go:512]: I stumbled over this until I dug into the logr changes and found that logr.Logger is now a struct - then this started to make sense.

A more complete commit message in this PR would have been useful.

File: klog.go [hunk 2]
```diff
@@ -698,7 +698,7 @@ func (buf *buffer) someDigits(i, d int) int {
 	return copy(buf.tmp[i:], buf.tmp[j:])
 }
 
-func (l *loggingT) println(s severity, logr logr.Logger, filter LogFilter, args ...interface{}) {
+func (l *loggingT) println(s severity, logr *logr.Logger, filter LogFilter, args ...interface{}) {
```
Comment (pohly) [inline klog.go:701]: Using "logr" as parameter name is awkward because is shadows the package name. The logr.go example uses "logger" as variable name - should we also use that here?

I understand that this PR is about making other changes, but as you seem to touch most of the lines where `logr` occurs, perhaps now would be a good opportunity to also change the name.

File: klog.go (hunk 1 above)
Comment (dims) [reply klog.go:512]: yep will do @pohly

File: klog.go (hunk 2 above)
Comment (dims) [reply klog.go]: on it.

File: klogr/klogr.go [hunk 3]
```diff
@@ -86,18 +91,31 @@ func copySlice(in []interface{}) []interface{} {
...
+//
+// This assumes that all logging paths are the same depth from the caller,
+// which should be a reasonable assumption since they are part of the same
+// interface.
 func framesToCaller() int {
+	// Figuring out the current depth is somewhat expensive. Saving the value
+	// amortizes most of that runtime cost.
+	if atomic.LoadInt32(&framesAtomic) != 0 {
+		return int(framesAtomic)
+	}
```
Comment (pohly) [inline klogr/klogr.go:113]: This enhancement is unrelated to the logr update, right? A separate PR for it might be better.

Regarding this function: I've not seen any case where it actually encountered an `autogeneratedFrameName`, i.e. the function always returned 1. I wonder whether it is still needed at all.

File: klog.go [hunk 4]
```diff
@@ -862,7 +862,7 @@ func (rb *redirectBuffer) Write(bytes []byte) (n int, err error) {
 // Use as:
 //   ...
 //   klog.SetLogger(zapr.NewLogger(zapLog))
-func SetLogger(logr logr.Logger) {
+func SetLogger(logr *logr.Logger) {
```
Comment (thockin) [inline klog.go:865]: I'm not convinced you need to take a pointer here.  Copying the logr.Logger struct should be cheap.  Then you can avoid this API break.

File: klogr/klogr.go (hunk 3 above)
Comment (thockin) [reply klogr/klogr.go:113]: This is a good point.  

Revisiting the original suggstion for this (https://groups.google.com/g/golang-nuts/c/AAcyf5muVPw) we no longer trigger the embedded-struct case in logr, so we probably CAN remove it!

File: klogr/klogr.go [hunk 5]
```diff
@@ -191,8 +209,8 @@ func pretty(value interface{}) string {
 	return strings.TrimSpace(string(buffer.Bytes()))
 }
 
-func (l klogger) Info(msg string, kvList ...interface{}) {
-	if l.Enabled() {
+func (l klogger) Info(level int, msg string, kvList ...interface{}) {
+	if l.Enabled(level) {
```
Comment (thockin) [inline klogr/klogr.go:213]: you should not need to check Enabled().  logr will do that and will only call this if true.

File: klogr/klogr.go [hunk 6]
```diff
@@ -236,35 +254,29 @@ func (l klogger) Error(err error, msg string, kvList ...interface{}) {
...
-	new.level = level
-	return new
-}
-
 // WithName returns a new logr.Logger with the specified name appended.  klogr
 // uses '/' characters to separate name elements.  Callers should not pass '/'
 // in the provided name string, but this library does not actually enforce that.
-func (l klogger) WithName(name string) logr.Logger {
+func (l klogger) WithName(name string) logr.LogSink {
 	new := l.clone()
```
Comment (thockin) [inline klogr/klogr.go:261]: you should not need a clone here - taking it by value-receiver already makes a copy.  The With* methods should be copy-on-write, so this is just wasted energy.  In glogr I got rid of clone() completely

File: klogr/klogr.go [hunk 7]
```diff
@@ -236,35 +254,29 @@ func (l klogger) Error(err error, msg string, kvList ...interface{}) {
...
 		new.prefix = l.prefix + "/"
 	}
 	new.prefix += name
-	return new
+	return &new
 }
 
-func (l klogger) WithValues(kvList ...interface{}) logr.Logger {
+func (l klogger) WithValues(kvList ...interface{}) logr.LogSink {
 	new := l.clone()
```
Comment (thockin) [inline klogr/klogr.go:270]: same

Comment (thockin) [reply klogr/klogr.go:270]: The trick to avoid clone here is:
```
// Three slice args forces a copy.
n := len(l.values)
l.values = append(l.values[:n:n], kvList...)
```

File: klogr/klogr.go [hunk 8]
```diff
@@ -236,35 +254,29 @@ func (l klogger) Error(err error, msg string, kvList ...interface{}) {
...
 	new.values = append(new.values, kvList...)
-	return new
+	return &new
 }
 
-func (l klogger) WithCallDepth(depth int) logr.Logger {
+func (l klogger) WithCallDepth(depth int) logr.LogSink {
 	new := l.clone()
```
Comment (thockin) [inline klogr/klogr.go:276]: same

File: klog.go (hunk 4 above)
Comment (pohly) [reply klog.go:865]: Code like https://github.com/kubernetes/klog/blob/69472594defe33aee4442f9266aadfddc89fe4b0/klog.go#L782-L785 must be able to check whether a logger has been set. That could also be done by checking for non-nil logging.logr.LogSink, though.

But API break is an interesting topic - isn't it also an API break if code using a logr.Logger interface (i.e. logr < 1.0.0-rc1) can no longer call SetLogger?

Comment (pohly) [reply klog.go:865]: Or do we argue that the semantic versioning of klog does not apply to those parts of it which use types from dependencies which themselves had no API stability guarantee yet?

Comment (thockin) [reply klog.go:865]: Ahh, well, you could internally store a pointer but accept a value.

Comment (pohly) [reply klog.go:865]: That's still an API change because previously valid values (interface) become invalid (now need the struct). The breakage is rather indirect because whatever code provided the logger implementation will also no longer compile (we cannot have logr v0.4 and v1.0.0 at the same time), but it's still breakage caused by updating klog.

Having said that, I prefer passing the logger by value. My  reasoning is that code which just passes logger instances around (formerly interfaces, now structs) continues to work unchanged.

File: klogr/klogr.go (hunk 5 above)
Comment (dims) [reply klogr/klogr.go:213]: on it.

File: klogr/klogr.go (hunk 3 above)
Comment (dims) [reply klogr/klogr.go:113]: ok removing the changes in this section

File: klog.go (hunk 4 above)
Comment (dims) [reply klog.go:865]: switching the signature back

File: klogr/klogr.go (hunk 6 above)
Comment (dims) [reply klogr/klogr.go:247]: Done!

File: klogr/klogr.go (hunk 7 above)
Comment (dims) [reply klogr/klogr.go:256]: Done!

File: klogr/klogr.go (hunk 8 above)
Comment (dims) [reply klogr/klogr.go:262]: Done!

**Comment by dims** [comment]
@thockin we still have other API breaks (after fixing up the `SetLogger`)

```
FAIL: k8s.io/klog/v2/klogr contains incompatible changes:
- klogger.Enabled: changed from func() bool to func(int) bool
- klogger.Info: changed from func(string, ...interface{}) to func(int, string, ...interface{})
- klogger.V: removed
- klogger.WithCallDepth: changed from func(int) github.com/go-logr/logr.Logger to func(int) github.com/go-logr/logr.LogSink
- klogger.WithName: changed from func(string) github.com/go-logr/logr.Logger to func(string) github.com/go-logr/logr.LogSink
- klogger.WithValues: changed from func(...interface{}) github.com/go-logr/logr.Logger to func(...interface{}) github.com/go-logr/logr.LogSink
```

**Comment by thockin** [comment]
None of those are user-facing APIs, though, right?   You should be able to make this change, re-vendor into k/k and recompile, I think?

**Comment by dims** [comment]
@thockin true.

File: klog.go (hunk 1 above)
Comment (RyouZhang) [reply klog.go:512]: It's not a good idea, a break changed

File: klog.go (hunk 4 above)
Comment (RyouZhang) [reply klog.go]: why not change logr.Logger to logr.LogSink ?

File: go.mod [hunk 9]
```diff
@@ -2,4 +2,4 @@ module k8s.io/klog/v2
 
 go 1.13
 
-require github.com/go-logr/logr v0.4.0
+require github.com/go-logr/logr v1.0.0-rc1
```
Comment (aojea) [inline go.mod:5]: you have v1.0.0 now https://github.com/go-logr/logr/releases/tag/v1.0.0

Comment (dims) [reply go.mod]: updated

File: klog.go (hunk 1 above)
Comment (dims) [reply klog.go:512]: we don't really have a choice here @RyouZhang

File: klogr/klogr.go [hunk 10]
```diff
@@ -68,24 +67,16 @@ type klogger struct {
...
-	return out
+func (l *klogger) Init(info logr.RuntimeInfo) {
+	l.callDepth += info.CallDepth
 }
 
 // Magic string for intermediate frames that we should ignore.
 const autogeneratedFrameName = "<autogenerated>"
 
+// Cached depth of this interface's log functions.
+var framesAtomic int32 // atomic
```
Comment (thockin) [inline klogr/klogr.go:78]: If you look at https://github.com/go-logr/glogr/commit/7a85ef23c9f6a08ad8444996c1d7250107b5e76c you will see that this logic id no longer needed since there isn't an interface in between.

**Review by thockin — COMMENTED** [review]
What discussion is needed?

File: klogr/klogr.go (hunk 10 above)
Comment (thockin) [reply klogr/klogr.go]: I meant that I think you don't need `autogeneratedFrameName` or `framesToCaller` or any of that :)

File: klog.go [hunk 11]
```diff
@@ -866,7 +866,14 @@ func SetLogger(logr logr.Logger) {
 	logging.mu.Lock()
```
Comment (serathius) [inline klog.go:866]: Wouldn't it make more sense to accept logr.LogSink interface and construct the logr.Logger instead of accepting the struct itself. 

I ask as I have seen multiple cases of users being confused with logr.Logger now being struct. Even though this change is purely for performance, and doesn't have any impact, some users misread this change. For example https://github.com/kubernetes/klog/pull/252

My first though about this change is that logr just splits client API (logr.Logger) from implementation interface (logr.LogSInk). Why then klog and would accept the client log.Logger instead of implementation logr.LogSink. 

It might be a subtle difference in API, but accepting interface makes it more clearer that intention of using logr is the same.

**Comment by thockin** [comment]
Look at it from an end-user point of view.  End users will never actually
hold a LogSink to pass in.  They will usually do something like:

```
logger := zapr.New(/* args... */)
klog.SetLogger(logger)
```

logr.Logger has no exported fields, so they can't reach in an extract the
LogSink.  Implementations like glogr, zapr, etc will not typically have a
direct LogSink constructor, just Logger.

If I understand correctly - the point of this capability is to let users
who use logr in their app send klog messages to the same place, right?  So
they are already going to have a logr.Logger.

File: klogr/klogr.go (hunk 10 above)
Comment (dims) [reply klogr/klogr.go]: whooooops! done.

File: klog.go (hunk 11 above)
Comment (dims) [reply klog.go:866]: response is https://github.com/kubernetes/klog/pull/242#issuecomment-888574114

**Comment by dims** [comment]
@thockin we have a verify script that fails :) so the question is do we disable that or we bump klog from v2 to v3

```
FAIL: k8s.io/klog/v2/klogr contains incompatible changes:
- klogger.Enabled: changed from func() bool to func(int) bool
- klogger.Info: changed from func(string, ...interface{}) to func(int, string, ...interface{})
- klogger.V: removed
- klogger.WithCallDepth: changed from func(int) github.com/go-logr/logr.Logger to func(int) github.com/go-logr/logr.LogSink
- klogger.WithName: changed from func(string) github.com/go-logr/logr.Logger to func(string) github.com/go-logr/logr.LogSink
- klogger.WithValues: changed from func(...interface{}) github.com/go-logr/logr.Logger to func(...interface{}) github.com/go-logr/logr.LogSink
```

**Comment by aojea** [comment]
> v2 to v3. we need to agree on how to roll this out. that's the discussion needed.

v3 looks easier to explain, no? I hit this today updating some dependencies that broke because of the logr changes, but I started to see more people doing `go get -u ...` that are starting to have problems ...

**Comment by thockin** [comment]
I see.  Yeah, if we are versioning klog then v3 is "most correct".  That said, are users really having problems that are not solved by a recompile?

**Comment by aojea** [comment]
> I see. Yeah, if we are versioning klog then v3 is "most correct". That said, are users really having problems that are not solved by a recompile?

yeah, this was mine today and I fixed using a replace statement, but you can similar ones on forums https://forum.golangbridge.org/t/go-lgr-compiler-error-despite-source-code-not-having-changed/23725
```
go get -u github.com/openshift/api                                  
# k8s.io/klog/v2
../../../../pkg/mod/k8s.io/klog/v2@v2.10.0/klog.go:705:10: invalid operation: logr != nil (mismatched types logr.Logger and nil)
../../../../pkg/mod/k8s.io/klog/v2@v2.10.0/klog.go:724:10: invalid operation: logr != nil (mismatched types logr.Logger and nil)
../../../../pkg/mod/k8s.io/klog/v2@v2.10.0/klog.go:742:10: invalid operation: logr != nil (mismatched types logr.Logger and nil)
../../../../pkg/mod/k8s.io/klog/v2@v2.10.0/klog.go:763:10: invalid operation: logr != nil (mismatched types logr.Logger and nil)
../../../../pkg/mod/k8s.io/klog/v2@v2.10.0/klog.go:782:11: invalid operation: loggr != nil (mismatched types logr.Logger and nil)
../../../../pkg/mod/k8s.io/klog/v2@v2.10.0/klog.go:783:3: undefined: logr.WithCallDepth
```

**Comment by thockin** [comment]
Yeah, the nil-ability of an interface is the one thing that has popped
up repeatedly and I don't have a great answer to it.

**Comment by pohly** [comment]
> > v2 to v3. we need to agree on how to roll this out. that's the discussion needed.
> 
> v3 looks easier to explain, no? I hit this today updating some dependencies that broke because of the logr changes, but I started to see more people doing `go get -u ...` that are starting to have problems ...

Doing a klog v3 will not avoid those problems, will it? `go get -u ...` will update logr to v1 but not klog from v2 to v3.

While I agree that klog v3 would be cleaner, I still think that sticking with v2 will be less pain overall. After a new minor release of klog, `go get -u` will produce a functional combination.

The rationale for that not being an API break in klog would be to retro-actively declare any klog API which depends on the unstable logr v0.4 API as "alpha".

**Comment by thockin** [comment]
we can pin klog v2 at 0.4.0 if we need, but that infects the whole
ecosystem, so moving to 1.0 is better.  AIUI converting the whole ecosystem
to v3 is equally challenging, of course

**Comment by pohly** [comment]
> we can pin klog v2 at 0.4.0 if we need, but that infects the whole
> ecosystem, so moving to 1.0 is better. 

Agreed. Even if we never update klog v2 itself to a logr > 0.4.0, there's no way to express in go.mod that updating to 1.0 breaks klog, so all downstream users will have to do that pinning themselves.

> AIUI converting the whole ecosystem to v3 is equally challenging

Equal to pinning v2 to logr 0.4.0 or equal to updating v2 to logr 1.0.0?

I think a migration to v3 is going to be a major pain and I am not even sure whether it can be done seamlessly once logr backends are involved. Compared to that, updating v2 to logr 1.0.0 seems easier to me.

**Comment by liggitt** [comment]
> so all downstream users will have to do that pinning themselves

and/or avoid running `go get -u` with incompatible `v0....` dependencies in their graph... that doesn't work well generally

**Comment by dims** [comment]
Longer discussion here: https://kubernetes.slack.com/archives/CHGFYJVAN/p1627926227001800

**Comment by dims** [comment]
Essentially,
- Anyone using `SetLogger` will have to cross check what they are using and will have to move to newer struct than the earlier interface
- A recompile should get folks back on their feet
- When we update k/k to this version of klog we may want to move to zapr (https://github.com/kubernetes/kubernetes/issues/104097)
- Hopefully there will be some guidance in a markdown in logr repo to help with transition ( cc @thockin )
- We should highlight this in kubernetes release notes for 1.23 (next release)
